File: typescript/src/evals/judge.ts
import { runAgent } from "../agents/runAgent.js";
import type { Harness } from "../harness/index.js";
import type { ChatModel } from "../providers/types.js";
import type { ToolRuntime } from "../tools/types.js";
import type { Tracer } from "../tracing/tracer.js";

export interface JudgeScores {
  accuracy: number;
  completeness: number;
  source_quality: number;
  process: number;
  disclosure: number;
}

export interface JudgeVerdict {
  scores: JudgeScores;
  justifications: Record<string, string>;
  flag_for_human_review: boolean;
}

export interface RunJudgeParams {
  task: string;
  response: string;
  traceSummary: string;
  model: ChatModel;
  harness: Harness;
  runtime: ToolRuntime;
  tracer: Tracer;
  evalId: string;
}

/** LLM-as-judge, single call, multi-criteria rubric (notes section 10; specs/prompts/judge.md). */
export async function runJudge(params: RunJudgeParams): Promise<JudgeVerdict> {
  const userPrompt = [
    `Original task: ${params.task}`,
    `Final response: ${params.response}`,
    `Trace summary: ${params.traceSummary}`,
  ].join("\n\n");

  const result = await runAgent({
    roleName: "judge",
    userPrompt,
    model: params.model,
    harness: params.harness,
    runtime: params.runtime,
    tracer: params.tracer,
    taskId: params.evalId,
    delegationDepth: 0,
  });

  try {
    return JSON.parse(result.text) as JudgeVerdict;
  } catch {
    throw new Error(`judge did not return valid JSON: ${result.text}`);
  }
}

File: python/src/multiagent_boilerplate/evals/judge.py
"""LLM-as-judge, single call, multi-criteria rubric (notes section 10; specs/prompts/judge.md)."""

from __future__ import annotations

import json

from ..agents.run_agent import RunAgentParams, run_agent
from ..harness import Harness
from ..providers.types import ChatModel
from ..tools.types import ToolRuntime
from ..tracing.tracer import Tracer


async def run_judge(
    task: str,
    response: str,
    trace_summary: str,
    model: ChatModel,
    harness: Harness,
    runtime: ToolRuntime,
    tracer: Tracer,
    eval_id: str,
) -> dict:
    user_prompt = "\n\n".join(
        [f"Original task: {task}", f"Final response: {response}", f"Trace summary: {trace_summary}"]
    )

    result = await run_agent(
        RunAgentParams(
            role_name="judge",
            user_prompt=user_prompt,
            model=model,
            harness=harness,
            runtime=runtime,
            tracer=tracer,
            task_id=eval_id,
            delegation_depth=0,
        )
    )

    try:
        return json.loads(result.text)
    except json.JSONDecodeError as error:
        raise ValueError(f"judge did not return valid JSON: {result.text}") from error
